apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PriorityHeap.cs

```diff
@@ -0,0 +1,168 @@
+                    //processing until after Add has completed.
+                    item.PriorityChanged += Item_PriorityChanged;
+
+                    _innerList.Insert(index, item);
+                }
+            }
+        }
+
+        public void Remove(T item)
+        {
+            lock (_innerList)
+            {
+                int index = 0;
+                //TODO: figure out a better way than a linear search here.
+                //Since this structure will mostly be used ~200 items or less,
+                //it will not be a major issue, but it would be nice to address it.
+                for (; index < _innerList.Count; index++)
+                {
+                    if (ReferenceEquals(_innerList[index], item))
+                    {
+                        break;
+                    }
+                }
+                if (index < _innerList.Count)
+                {
+                    _innerList.RemoveAt(index);
+                    item.PriorityChanged -= Item_PriorityChanged;
+                }
+            }
+        }
+
+        public void RemoveLast()
+        {
+            lock (_innerList)
+            {
+                if (_innerList.Count > 0)
+                {
+                    var item = _innerList[_innerList.Count - 1];
+                    item.PriorityChanged -= Item_PriorityChanged;
+                    _innerList.RemoveAt(_innerList.Count - 1);
+                }
+            }
+        }
+
+        private void Item_PriorityChanged(object sender, EventArgs e)
+        {
+            lock (_innerList)
+            {
+                T item = (T) sender;
+                lock (_innerList)
+                {
+                    //Re-Add the item sorted.
+                    Remove(item);
+                    Add(item);
+                }
+            }
+        }
+
+        private bool TryFindIndex(T itemToFind, out int index)
+        {
+            lock (_innerList)
+            {
+                index = TryFindIndex(itemToFind, 0, _innerList.Count - 1);
+                return index < _innerList.Count && ReferenceEquals(_innerList[index], itemToFind);
+            }
+        }
+
+        private int TryFindIndex(T itemToFind, int start, int end)
+        {
+            if (start > end)
+            {
+                return start;
+            }
+            int mid = (start + end) / 2;
+            var comparison = itemToFind.CompareTo(_innerList[mid]);
+            if (comparison == 0)
+            {
+                return mid;
+            }
+            else if (comparison > 0)
+            {
+                return TryFindIndex(itemToFind, mid + 1, end);
+            }
+            else
+            {
+                return TryFindIndex(itemToFind, start, mid - 1);
+            }
+        }
+
+        public T GetTop()
+        {
+            lock (_innerList)
+            {
+                return _innerList.Count > 0 ? _innerList[0] : default(T);
```

Comment:
I reviewed the use cases and even in session pool I do not think its possible for count to be zero here.  So I believe invalidoperationexception would be fine right now.  My main concern is if the pooling code changes slightly.  I was always ok with the idea of a low lock design where I added an entry to the pool, checked to see if it was too big and then removed an item.  It has drawbacks in that the pool can become (very temporarily) bigger than max size.  It may also reduce an amount more than necessary due to concurrency -- but it makes it pretty simple to manage.
However if this ever changes to a design that has stricter guarantees, the count may drop to zero and I think its fine for the prioritylist to return null in that case and let the caller figure out what to do.